clang/test/SemaSYCL/union-kernel-param-neg.cpp

```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -fcxx-exceptions -verify -fsyntax-only %s
+
+// This test checks if compiler reports compilation error on an attempt to pass
+// accessor/sampler/pointer as SYCL kernel parameter inside union.
+
+#include "sycl.hpp"
+using namespace cl::sycl;
+
+union union_with_sampler {
+  cl::sycl::sampler smpl;
+  // expected-error@-1 {{'cl::sycl::sampler' cannot be used as the type of a kernel parameter}}
+};
+
+union union_with_pointer {
+  int *ptr_in_union;
+  // expected-error@-1 {{'int *' cannot be used as the type of a kernel parameter}}
```

Comment:
Should we really prohibit using of pointers inside unions?
SYCL spec doesn't say about it.

Reply:
My understanding is pointers need to be wrapped separately, don't they?  I thought that based on what @elizabethandrews  has been working on lately.

We need to make sure we separately wrap them/convert them to globals, which we cannot do inside a union.

Reply:
I think @elizabethandrews is wrapping pointers to avoid presence of raw pointer in kernel arguments, because in this case they are checked by underlying runtime and if they are not initialized, the RT will give an error. That is why structs decomposition caused regressions for cases when some pointer was wrapped by a struct but and it wasn't used as USM pointer. Struct decomposition also helped to fix performance problems caused by the fact that wrapped pointers were passed with generic address space instead of global.
So, we cannot decompose unions, that is why USM pointers passed inside unions likely will be always passed with generic address space, but it will be only performance problem, not a functional one. So, while USM extension spec doesn't prohibit passing USM pointers inside unions (I feel that it doesn't but I'm not really an expert in it, probably @jbrodman can help here), we can emit a warning/note that will say that USM pointer won't give high performance here, because technically SYCL core spec doesn't prohibit such code.

Reply:
In the latest patch I'm only handling pointers which are inside structs. Wrapping isn't done for pointers which are passed directly as a kernel argument since this seems to mess with USM stuff. 

Since union is handled as a single non-decomposed type I think pointers inside will just be ignored for the 'wrapping' (which I guess is fine?). Changing the address space to global however will not happen. Since that requires decomposition.

Reply:
I see!  Thanks for the explaination!  @smanna12 : it sounds that we can just remove the limit on pointers.  

Thanks you two!

Reply:
By 'pointers inside structs' I mean pointers which were originally fields of a struct type kernel argument (which we then decomposed and passed individual fields for.)

Reply:
I have removed the limitation on pointers inside union. Thanks everyone for the explanation.